# 🔗 **Data Merging and Feature Engineering**

This notebook brings together all 8 cleaned datasets into a single
unified master dataframe and engineers three additional predictive
features. This master dataset is the direct input to all modelling
notebooks.

## Notebook Structure
1. Read in all 8 cleaned datasets
2. Standardise date formats across all datasets
3. Merge all datasets onto the food price base
4. Handle remaining missing values
5. Engineer price lag features
6. Engineer harvest season indicator
7. Engineer related commodity average price feature
8. Save final master dataset

###**MOUNTING MY DRIVE**

In [ ]:
#connect my drive to colab
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


### **IMPORTED LIBRARIES/DEFINED FUNCTIONS**

In [ ]:
#importing libraries
import pandas as pd
import numpy as np

In [ ]:
# --- defined function ---
#for reading in dataset
def read_dataset_csv(path):
  df = path
  pd.read_csv(df).head
  return pd.read_csv(df)

def wrangle (df):
  #checking shape and structure of the data
  print('--- SHAPE ---')
  print(df.shape)

  #checking the info
  print('--- INFO ---')
  print(df.info())

  #checking for duplicate rows
  print('--- DUPLICATE ROWS ---')
  print(df.duplicated().sum())

  #checking for  null/missing values
  print('--- NULLS ---')
  print(df.isnull().sum())

### **READING IN THE CLEANED DATASETS**

In [ ]:
#reading the commodity_price dataset
commodity_price = read_dataset_csv('/content/drive/MyDrive/price forecasting project data(cleaned)/wfp_food_price_clean.csv')
commodity_price.head()

,date,state,commodity,price
0,2017-01-15,Abia,Maize (yellow),177.12
1,2017-01-15,Abia,Rice (imported),1450.00
2,2017-01-15,Abia,Rice (local),1073.00
3,2017-01-15,Abia,Yam,504.00
4,2017-01-15,Borno,Beans (white),783.50


In [ ]:
#overview of the dataset
wrangle(commodity_price)

--- SHAPE ---
(2846, 4)
--- INFO ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2846 entries, 0 to 2845
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   date       2846 non-null   object 
 1   state      2846 non-null   object 
 2   commodity  2846 non-null   object 
 3   price      2846 non-null   float64
dtypes: float64(1), object(3)
memory usage: 89.1+ KB
None
--- DUPLICATE ROWS ---
0
--- NULLS ---
date         0
state        0
commodity    0
price        0
dtype: int64


In [ ]:
#reading the inflation_rate dataset
inflation_rate = read_dataset_csv('/content/drive/MyDrive/price forecasting project data(cleaned)/inflation_rate_clean.csv')
inflation_rate.head()

,date,inflation_rate
0,2017-01-01,23.746418
1,2017-02-01,24.218748
2,2017-03-01,24.753596
3,2017-04-01,25.258312
4,2017-05-01,25.898688


In [ ]:
#overview of the dataset
wrangle(inflation_rate)

--- SHAPE ---
(96, 2)
--- INFO ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96 entries, 0 to 95
Data columns (total 2 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   date            96 non-null     object 
 1   inflation_rate  96 non-null     float64
dtypes: float64(1), object(1)
memory usage: 1.6+ KB
None
--- DUPLICATE ROWS ---
0
--- NULLS ---
date              0
inflation_rate    0
dtype: int64


In [ ]:
#reading in exchange_rate dataset
exchange_rate = read_dataset_csv('/content/drive/MyDrive/price forecasting project data(cleaned)/exchange_rate_clean.csv')
exchange_rate.head()

,exchange_rate,date
0,305.202381,2017-01-01
1,305.777273,2018-01-01
2,306.845455,2019-01-01
3,306.959091,2020-01-01
4,381.000000,2021-01-01


In [ ]:
#overview of the dataset
wrangle(exchange_rate)

--- SHAPE ---
(96, 2)
--- INFO ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96 entries, 0 to 95
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   exchange_rate  96 non-null     float64
 1   date           96 non-null     object 
dtypes: float64(1), object(1)
memory usage: 1.6+ KB
None
--- DUPLICATE ROWS ---
0
--- NULLS ---
exchange_rate    0
date             0
dtype: int64


In [ ]:
#reading in fuel_price dataset
fuel_price = read_dataset_csv('/content/drive/MyDrive/price forecasting project data(cleaned)/pms_clean_final.csv')
fuel_price.head()

,state,date,pms_price
0,Abia,2017-03-01,150.4
1,Abia,2017-04-01,149.0
2,Abia,2017-05-01,151.9
3,Abia,2017-06-01,149.7
4,Abia,2017-07-01,147.0


In [ ]:
#overview of the dataset
wrangle(fuel_price)

--- SHAPE ---
(1274, 3)
--- INFO ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1274 entries, 0 to 1273
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   state      1274 non-null   object 
 1   date       1274 non-null   object 
 2   pms_price  1274 non-null   float64
dtypes: float64(1), object(2)
memory usage: 30.0+ KB
None
--- DUPLICATE ROWS ---
0
--- NULLS ---
state        0
date         0
pms_price    0
dtype: int64


In [ ]:
#reading in mpr dataset
mpr = read_dataset_csv('/content/drive/MyDrive/price forecasting project data(cleaned)/mpr_clean.csv')
mpr.head()

,date,mpr
0,2017-01-01,14.0
1,2017-02-01,14.0
2,2017-03-01,14.0
3,2017-04-01,14.0
4,2017-05-01,14.0


In [ ]:
#overview of the dataset
wrangle(mpr)

--- SHAPE ---
(96, 2)
--- INFO ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96 entries, 0 to 95
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   date    96 non-null     object 
 1   mpr     96 non-null     float64
dtypes: float64(1), object(1)
memory usage: 1.6+ KB
None
--- DUPLICATE ROWS ---
0
--- NULLS ---
date    0
mpr     0
dtype: int64


In [ ]:
#reading in the rainfall dataset
rainfall = read_dataset_csv('/content/drive/MyDrive/price forecasting project data(cleaned)/rainfall_clean.csv')
rainfall.head()

,date,state,rainfall
0,2017-04-01,Abia,0.0
1,2017-05-01,Abia,0.0
2,2017-06-01,Abia,0.0
3,2017-07-01,Abia,0.0
4,2017-08-01,Abia,0.0


In [ ]:
#overview of the dataset
wrangle(rainfall)

--- SHAPE ---
(1302, 3)
--- INFO ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1302 entries, 0 to 1301
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   date      1302 non-null   object 
 1   state     1302 non-null   object 
 2   rainfall  1302 non-null   float64
dtypes: float64(1), object(2)
memory usage: 30.6+ KB
None
--- DUPLICATE ROWS ---
0
--- NULLS ---
date        0
state       0
rainfall    0
dtype: int64


In [ ]:
#reading in the food import volume dataset
fiv = read_dataset_csv('/content/drive/MyDrive/price forecasting project data(cleaned)/fiv_clean.csv')
fiv.head()

,date,commodity,import_volume
0,2017-01-01,Beans (white),599.0
1,2017-02-01,Beans (white),599.0
2,2017-03-01,Beans (white),599.0
3,2017-04-01,Beans (white),599.0
4,2017-05-01,Beans (white),599.0


In [ ]:
#overview of the dataset
wrangle(fiv)

--- SHAPE ---
(552, 3)
--- INFO ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 552 entries, 0 to 551
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   date           552 non-null    object 
 1   commodity      552 non-null    object 
 2   import_volume  552 non-null    float64
dtypes: float64(1), object(2)
memory usage: 13.1+ KB
None
--- DUPLICATE ROWS ---
0
--- NULLS ---
date             0
commodity        0
import_volume    0
dtype: int64


In [ ]:
#reading in the conflict index dataset
conflict_index = read_dataset_csv('/content/drive/MyDrive/price forecasting project data(cleaned)/conflict_index_monthly.csv')
conflict_index.head()

,state,date,total_events,conflict_score,fatalities
0,Abia,2017-02-01,1,1,0
1,Abia,2017-03-01,5,15,2
2,Abia,2017-04-01,2,2,0
3,Abia,2017-05-01,1,1,0
4,Abia,2017-06-01,1,5,3


In [ ]:
#overview of the dataset
wrangle(conflict_index)

--- SHAPE ---
(1150, 5)
--- INFO ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1150 entries, 0 to 1149
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   state           1150 non-null   object
 1   date            1150 non-null   object
 2   total_events    1150 non-null   int64 
 3   conflict_score  1150 non-null   int64 
 4   fatalities      1150 non-null   int64 
dtypes: int64(3), object(2)
memory usage: 45.1+ KB
None
--- DUPLICATE ROWS ---
0
--- NULLS ---
state             0
date              0
total_events      0
conflict_score    0
fatalities        0
dtype: int64


## **🔀 Merging All 8 Datasets**

### Merge Strategy

The food price dataset serves as the base — all other features
are joined onto it. The merge key varies by dataset type:

| Dataset | Merge Key | Reason |
|---------|-----------|--------|
| Rainfall | date + state | State-level monthly data |
| PMS Petrol Price | date + state | State-level monthly data |
| Conflict Index | date + state | State-level monthly data |
| Inflation Rate | date only | National figure — same for all states |
| Exchange Rate | date only | National figure — same for all states |
| MPR | date only | National figure — same for all states |
| Import Volume | date + commodity | Commodity-level annual data |

All merges use **how='left'** to preserve every row in the
food price base dataset. Unmatched rows receive NaN which
are handled in the missing value treatment step.

### Date Standardisation
Before merging, all date columns were converted from object
dtype to pandas Period format (YYYY-MM). This was necessary
because the WFP food price dataset recorded dates on the
15th of each month while all other datasets used the 1st.
Converting to year-month period removes the day component
entirely, ensuring January 2017 in one dataset correctly
matches January 2017 in all others.

### Missing Value Treatment
After merging, three columns required targeted imputation:

- **rainfall_mm** — 42 nulls — forward filled then backward
  filled by state. Rainfall patterns are stable enough
  month-to-month that the nearest known value is a
  reasonable estimate
- **pms_price_ngn** — 124 nulls — forward filled then
  backward filled by state. Nigerian petrol prices
  remain fixed for extended periods making forward
  fill highly appropriate
- **import_volume_tonnes** — 1,132 nulls — forward filled
  by commodity then remaining nulls filled with zero.
  Zero is the correct value for commodities with no
  recorded import activity

### **MERGING OF DATASETS**

In [ ]:
# convert all dates to datetime then to monthly period
# ensures all datasets join on the same date format
for df in [commodity_price, inflation_rate, exchange_rate, mpr, rainfall, fuel_price, fiv, conflict_index]:
    df['date'] = pd.to_datetime(df['date'])

for df in [commodity_price, inflation_rate, exchange_rate, mpr, rainfall, fuel_price, fiv, conflict_index]:
    df['date'] = pd.to_datetime(df['date']).dt.to_period('M')

# base dataset -- all other features joined onto food prices
master = commodity_price.copy()

# state + date join -- rainfall varies by state and month
master = master.merge(rainfall, on=['date', 'state'], how='left')
master = master.merge(fuel_price, on=['date', 'state'], how='left')
master = master.merge(conflict_index, on=['date', 'state'], how='left')

master.head()


,date,state,commodity,price,rainfall,pms_price,total_events,conflict_score,fatalities
0,2017-01,Abia,Maize (yellow),177.12,0.0,NaN,NaN,NaN,NaN
1,2017-01,Abia,Rice (imported),1450.00,0.0,NaN,NaN,NaN,NaN
2,2017-01,Abia,Rice (local),1073.00,0.0,NaN,NaN,NaN,NaN
3,2017-01,Abia,Yam,504.00,0.0,NaN,NaN,NaN,NaN
4,2017-01,Borno,Beans (white),783.50,0.0,NaN,35.0,159.0,446.0


In [ ]:
# national date-only joins -- these features are the same
# value for all states in a given month
master = master.merge(inflation_rate, on='date', how='left')
master = master.merge(exchange_rate, on='date', how='left')
master = master.merge(mpr, on='date', how='left')

master.head()

,date,state,commodity,price,rainfall,pms_price,total_events,conflict_score,fatalities,inflation_rate,exchange_rate,mpr
0,2017-01,Abia,Maize (yellow),177.12,0.0,NaN,NaN,NaN,NaN,23.746418,305.202381,14.0
1,2017-01,Abia,Rice (imported),1450.00,0.0,NaN,NaN,NaN,NaN,23.746418,305.202381,14.0
2,2017-01,Abia,Rice (local),1073.00,0.0,NaN,NaN,NaN,NaN,23.746418,305.202381,14.0
3,2017-01,Abia,Yam,504.00,0.0,NaN,NaN,NaN,NaN,23.746418,305.202381,14.0
4,2017-01,Borno,Beans (white),783.50,0.0,NaN,35.0,159.0,446.0,23.746418,305.202381,14.0


In [ ]:
#merge date + commodity datasets
master = master.merge(fiv, on=['date', 'commodity'], how='left')

# conflict nulls mean zero recorded events that month
master[['total_events', 'conflict_score', 'fatalities']] = master[['total_events', 'conflict_score', 'fatalities']].fillna(0)


In [ ]:
#check
wrangle(master)

--- SHAPE ---
(2846, 13)
--- INFO ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2846 entries, 0 to 2845
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype    
---  ------          --------------  -----    
 0   date            2846 non-null   period[M]
 1   state           2846 non-null   object   
 2   commodity       2846 non-null   object   
 3   price           2846 non-null   float64  
 4   rainfall        2816 non-null   float64  
 5   pms_price       2744 non-null   float64  
 6   total_events    2846 non-null   float64  
 7   conflict_score  2846 non-null   float64  
 8   fatalities      2846 non-null   float64  
 9   inflation_rate  2846 non-null   float64  
 10  exchange_rate   2846 non-null   float64  
 11  mpr             2846 non-null   float64  
 12  import_volume   2091 non-null   float64  
dtypes: float64(10), object(2), period[M](1)
memory usage: 289.2+ KB
None
--- DUPLICATE ROWS ---
0
--- NULLS ---
date                0
state          

In [ ]:
#filling null values
# fill rainfall and pms with forward fill then backward fill
master['rainfall'] = master.groupby('state')['rainfall'].ffill().bfill()
master['pms_price'] = master.groupby('state')['pms_price'].ffill().bfill()

# fill import volume with forward fill by commodity then fill remaining with 0
master['import_volume'] = master.groupby('commodity')['import_volume'].ffill().bfill().fillna(0)

In [ ]:
#renaming the columns
master = master.rename(columns={
    'price': 'price_ngn',
    'rainfall': 'rainfall_mm',
    'pms_price': 'pms_price_ngn',
    'import_volume': 'import_volume_tonnes',
    'inflation_rate': 'inflation_rate_pct',
    'exchange_rate': 'exchange_rate_ngn_usd',
    'mpr': 'mpr_pct',
    'total_events': 'conflict_events',
    'conflict_score': 'conflict_score_weighted',
    'fatalities': 'conflict_fatalities'
})

In [ ]:
#check
wrangle(master)

--- SHAPE ---
(2846, 13)
--- INFO ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2846 entries, 0 to 2845
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype    
---  ------                   --------------  -----    
 0   date                     2846 non-null   period[M]
 1   state                    2846 non-null   object   
 2   commodity                2846 non-null   object   
 3   price_ngn                2846 non-null   float64  
 4   rainfall_mm              2846 non-null   float64  
 5   pms_price_ngn            2846 non-null   float64  
 6   conflict_events          2846 non-null   float64  
 7   conflict_score_weighted  2846 non-null   float64  
 8   conflict_fatalities      2846 non-null   float64  
 9   inflation_rate_pct       2846 non-null   float64  
 10  exchange_rate_ngn_usd    2846 non-null   float64  
 11  mpr_pct                  2846 non-null   float64  
 12  import_volume_tonnes     2846 non-null   float64  
dtypes: float64

In [ ]:
#master.head()
master['commodity'].unique()

array(['Maize (yellow)', 'Rice (imported)', 'Rice (local)', 'Yam',
       'Beans (white)', 'Tomatoes', 'Maize (white)'], dtype=object)

In [ ]:
master.to_csv('/content/drive/MyDrive/price forecasting project data(cleaned)/master_dataset_clean.csv', index=False)

## **⚙️ Feature Engineering**

Three additional features were engineered from the master
dataset to improve the model's predictive power beyond
what raw historical data alone can provide.

---

### Feature 1 — Price Lag Features

**Concept:** Food prices are highly autocorrelated — this
month's price is strongly predicted by last month's price.
Providing the model with explicit access to recent price
history allows it to learn these momentum patterns directly.

**Implementation:** Three lag features were created for
each state-commodity combination:
- **price_lag1** — retail price from 1 month ago
- **price_lag2** — retail price from 2 months ago
- **price_lag3** — retail price from 3 months ago

Data was sorted by state, commodity and date before
applying shift() to ensure lags were calculated within
the correct group. The first 1, 2 and 3 months of each
state-commodity series produce NaN values — this is
expected and represents the warmup period where
insufficient history exists. These rows are dropped
before model training.

---

### Feature 2 — Harvest Season Indicator

**Concept:** Nigerian agricultural commodities follow
predictable annual harvest cycles. During harvest months
supply increases dramatically and prices typically fall.
A binary indicator tells the model whether a given
month is a harvest month for a given commodity.

**Harvest month definitions** based on Nigerian
agricultural calendar:

| Commodity | Harvest Months |
|-----------|---------------|
| Maize (white) | August, September, October |
| Maize (yellow) | August, September, October |
| Rice (imported) | October, November, December |
| Rice (local) | October, November, December |
| Beans (white) | October, November |
| Yam | August, September, October |
| Tomatoes | November, December, January |

**Implementation:** A dictionary mapping each commodity
to its harvest months was defined. A lambda function
applied row-by-row checks whether the month of each
row falls within the harvest months for that row's
commodity and assigns 1 (harvest) or 0 (non-harvest).

**Result:** 727 harvest month rows (22.5%) and 2,119
non-harvest rows (77.5%) — a realistic distribution
for a 12-month cycle with 2-3 harvest months per
commodity.

---

### Feature 3 — Related Commodity Average Price

**Concept:** Food commodity prices influence each other.
When Maize becomes expensive, demand shifts to Beans
as a cheaper protein source — pushing Beans prices up.
Farmers also make planting decisions based on relative
commodity profitability. The average price of other
commodities in the same state and month captures this
cross-commodity price pressure.

**Implementation:** The mean price across all commodities
was computed for each state-month combination and merged
back onto the master dataset. Each row receives the
average price of all commodities in its state that month
as a single summary signal of overall food price pressure.

---

### Final Master Dataset

| Property | Value |
|----------|-------|
| Shape | 2,846 rows ×  18 columns |
| Date range | 2017 to 2024 |
| States | 13 |
| Commodities | 5 |
| Features | 13 original + 3 engineered |
| Missing values | 0 |
| Saved as | master_dataset_clean.csv |

**Column reference:**

| Column | Description | Unit |
|--------|-------------|------|
| date | Year-month period | YYYY-MM |
| state | Nigerian state name | — |
| commodity | Food commodity name | — |
| price_ngn | Monthly average retail price | Nigerian Naira |
| rainfall_mm | Monthly rainfall | Millimetres |
| pms_price_ngn | Monthly petrol pump price | Nigerian Naira |
| total_events | Raw conflict event count | Count |
| conflict_score | Severity-weighted conflict score | Weighted count |
| fatalities | Conflict fatalities | Count |
| inflation_rate_pct | Monthly food inflation rate | Percentage |
| exchange_rate_ngn_usd | NGN per USD exchange rate | Naira per Dollar |
| mpr_pct | Central Bank policy rate | Percentage |
| import_volume_tonnes | Monthly food import estimate | Metric tonnes |
| price_lag1 | Retail price 1 month ago | Nigerian Naira |
| price_lag2 | Retail price 2 months ago | Nigerian Naira |
| price_lag3 | Retail price 3 months ago | Nigerian Naira |
| harvest_season | Harvest month indicator | Binary (0 or 1) |
| other_commodity_avg_price | Average price of other commodities in same state/month | Nigerian Naira |

### **FEATURE ENGINEERING**

In [ ]:
#reading in the dataset
master = pd.read_csv('/content/drive/MyDrive/price forecasting project data(cleaned)/master_dataset_clean.csv')

#converting the dtype of date from object to datetime
master['date'] = pd.to_datetime(master['date'])
print(master.shape)
print(master.columns.tolist())

(2846, 13)
['date', 'state', 'commodity', 'price_ngn', 'rainfall_mm', 'pms_price_ngn', 'conflict_events', 'conflict_score_weighted', 'conflict_fatalities', 'inflation_rate_pct', 'exchange_rate_ngn_usd', 'mpr_pct', 'import_volume_tonnes']


In [ ]:
# sort first so lags are calculated in correct time order
master = master.sort_values(['state', 'commodity', 'date']).reset_index(drop=True)

# create lag features
master['price_lag1'] = master.groupby(['state', 'commodity'])['price_ngn'].shift(1)
master['price_lag2'] = master.groupby(['state', 'commodity'])['price_ngn'].shift(2)
master['price_lag3'] = master.groupby(['state', 'commodity'])['price_ngn'].shift(3)

In [ ]:
master.head()

,date,state,commodity,price_ngn,rainfall_mm,pms_price_ngn,conflict_events,conflict_score_weighted,conflict_fatalities,inflation_rate_pct,exchange_rate_ngn_usd,mpr_pct,import_volume_tonnes,price_lag1,price_lag2,price_lag3
0,2021-03-01,Abia,Maize (white),280.52,0.0,158.57,8.0,33.0,6.0,46.082248,381.000000,11.5,98313.0,NaN,NaN,NaN
1,2021-05-01,Abia,Maize (white),345.27,0.0,166.13,13.0,47.0,8.0,47.029667,400.126667,11.5,98313.0,280.52,NaN,NaN
2,2021-07-01,Abia,Maize (white),400.30,0.0,182.23,1.0,1.0,0.0,47.961686,410.121000,11.5,98313.0,345.27,280.52,NaN
3,2021-08-01,Abia,Maize (white),294.04,0.0,171.51,7.0,25.0,2.0,48.470936,410.385455,11.5,98313.0,400.30,345.27,280.52
4,2021-10-01,Abia,Maize (white),333.88,0.0,170.25,9.0,33.0,6.0,49.526367,411.247368,11.5,98313.0,294.04,400.30,345.27


In [ ]:
wrangle(master)

--- SHAPE ---
(2846, 16)
--- INFO ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2846 entries, 0 to 2845
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   date                     2846 non-null   datetime64[ns]
 1   state                    2846 non-null   object        
 2   commodity                2846 non-null   object        
 3   price_ngn                2846 non-null   float64       
 4   rainfall_mm              2846 non-null   float64       
 5   pms_price_ngn            2846 non-null   float64       
 6   conflict_events          2846 non-null   float64       
 7   conflict_score_weighted  2846 non-null   float64       
 8   conflict_fatalities      2846 non-null   float64       
 9   inflation_rate_pct       2846 non-null   float64       
 10  exchange_rate_ngn_usd    2846 non-null   float64       
 11  mpr_pct                  2846 non-null   float64       
 

In [ ]:
#define harvest months per commodity
harvest_months = {
    'Maize (white)': [8, 9, 10],
    'Maize (yellow)': [8, 9, 10],
    'Rice (imported)': [10, 11, 12],
    'Rice (local)': [10, 11, 12],
    'Beans (white)': [10, 11],
    'Yam': [8, 9, 10],
    'Tomatoes': [11, 12, 1]
}

#extract month number from date
master['month'] = master['date'].dt.month

#create harvest season indicator
master['harvest_season'] = master.apply(
    lambda row: 1 if row['month'] in harvest_months.get(row['commodity'], []) else 0,
    axis=1
)

#drop the temporary month column
master = master.drop(columns=['month'])

#check
print(master['harvest_season'].value_counts())

harvest_season
0    2119
1     727
Name: count, dtype: int64


In [ ]:
# calculate average price of OTHER commodities per state per month
# IMPORTANT: each commodity is excluded from its own average
# to prevent mild target leakage where a commodity's price
# appears in its own predictor feature

other_commodity_avg = (
    master.groupby(['date', 'state', 'commodity'])
    .apply(lambda grp: (
        master[
            (master['date']      == grp['date'].iloc[0]) &
            (master['state']     == grp['state'].iloc[0]) &
            (master['commodity'] != grp['commodity'].iloc[0])
        ]['price_ngn'].mean()
    ))
    .reset_index()
    .rename(columns={0: 'other_commodity_avg_price'})
)

master = master.merge(
    other_commodity_avg,
    on=['date', 'state', 'commodity'],
    how='left'
)

# confirm no nulls introduced
print(f'Nulls in other_commodity_avg_price: '
      f'{master["other_commodity_avg_price"].isnull().sum()}')

Nulls in other_commodity_avg_price: 157


/tmp/ipykernel_214/636433714.py:8: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  .apply(lambda grp: (


In [ ]:
master.head()


,date,state,commodity,price_ngn,rainfall_mm,pms_price_ngn,conflict_events,conflict_score_weighted,conflict_fatalities,inflation_rate_pct,exchange_rate_ngn_usd,mpr_pct,import_volume_tonnes,price_lag1,price_lag2,price_lag3,harvest_season,other_commodity_avg_price
0,2021-03-01,Abia,Maize (white),280.52,0.0,158.57,8.0,33.0,6.0,46.082248,381.000000,11.5,98313.0,NaN,NaN,NaN,0,786.520
1,2021-05-01,Abia,Maize (white),345.27,0.0,166.13,13.0,47.0,8.0,47.029667,400.126667,11.5,98313.0,280.52,NaN,NaN,0,887.875
2,2021-07-01,Abia,Maize (white),400.30,0.0,182.23,1.0,1.0,0.0,47.961686,410.121000,11.5,98313.0,345.27,280.52,NaN,0,1038.740
3,2021-08-01,Abia,Maize (white),294.04,0.0,171.51,7.0,25.0,2.0,48.470936,410.385455,11.5,98313.0,400.30,345.27,280.52,1,940.220
4,2021-10-01,Abia,Maize (white),333.88,0.0,170.25,9.0,33.0,6.0,49.526367,411.247368,11.5,98313.0,294.04,400.30,345.27,1,1030.510


In [ ]:
wrangle(master)

--- SHAPE ---
(2846, 18)
--- INFO ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2846 entries, 0 to 2845
Data columns (total 18 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   date                       2846 non-null   datetime64[ns]
 1   state                      2846 non-null   object        
 2   commodity                  2846 non-null   object        
 3   price_ngn                  2846 non-null   float64       
 4   rainfall_mm                2846 non-null   float64       
 5   pms_price_ngn              2846 non-null   float64       
 6   conflict_events            2846 non-null   float64       
 7   conflict_score_weighted    2846 non-null   float64       
 8   conflict_fatalities        2846 non-null   float64       
 9   inflation_rate_pct         2846 non-null   float64       
 10  exchange_rate_ngn_usd      2846 non-null   float64       
 11  mpr_pct                    2846

In [ ]:
# save final master dataset with all engineered features
# filename must match exactly what downstream notebooks load
master.to_csv(
    '/content/drive/MyDrive/price forecasting '
    'project data(cleaned)/'
    'master_dataset_clean.csv',
    index=False
)

print(f'Master dataset saved successfully.')
print(f'Shape: {master.shape}')
print(f'Columns: {master.columns.tolist()}')

Master dataset saved successfully.
Shape: (2846, 18)
Columns: ['date', 'state', 'commodity', 'price_ngn', 'rainfall_mm', 'pms_price_ngn', 'conflict_events', 'conflict_score_weighted', 'conflict_fatalities', 'inflation_rate_pct', 'exchange_rate_ngn_usd', 'mpr_pct', 'import_volume_tonnes', 'price_lag1', 'price_lag2', 'price_lag3', 'harvest_season', 'other_commodity_avg_price']
